### PR TITLE
[native pos] Change from Operator to Task Level Adapter

### PR DESCRIPTION
**Pull request by shrinidhijoshi**
## Summary

This PR introduces below changes

- Add  - `PrestoSparkNativeTaskExecutorFactory` singleton class (executor lifetime). This class is responsible for executing `PrestoSparkNativeTaskRDD`s.  
- Remove - `NativeExecutionNode`, `NativeExecutionOperator` , `NativeExecutionInfo` and corresponding plan rewrites and special handling
- Rework Spark Metrics Integration for new adapter

## Design 
image

The new  `PrestoSparkNativeTaskExecutorFactory` does below steps
- Create the CPP process once when the first task arrives. 
- For every task that arrives
    - Forward the task descriptors to the CPP process for execution. as `BatchTaskUpdateInfo` object
    - Return a Result `Iterator` to Spark layer, which
        - Wait on `page`s or taskCompletion from the CPP process
        - Handle any exceptions during task processing
        - If task is complete, populate `TaskInfoAccumulator `to propagate statistics 

## Implementation notes/caveats
The current implementation does not introduce any new quirks / caveats. Rather exposes some from the old adapter
- Piggyback on the existing classes for process management (`NativeExecutionProcess`, `NativeExecutionTask`, `NativeExecutionProcessFactory`, etc)
- Usage of a `dummyRemoteSourceTaskId` when generating shuffle read splits. More details in comments

## Test Plan
- [x] Unit test 
    - [x] All existing unit tests provide coverage over the newly added code
- [ ] E2E tests / Manual verification
    - [x] Run sample queries in prod 
        - [20230608_202250_00000_seeaq](https://www.internalfb.com/intern/presto/query/?query_id=20230608_202250_00000_seeaq) 
    - [x] Run sample queries in verifier
- [ ] Metrics Integration
    - [x] PrestoLookup Tool 
    - [x] Spark datasets
- [ ] Dev Tooling/Monitoring/Debugging compatibility
    - [x] 20230612_192556_00000_e6ttx
    - [ ] Test LocalMode
    - [ ] Test LocalMode with debugger
  
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java [hunk 1]
```diff
@@ -144,7 +144,7 @@ public void stop()
     public Optional<TaskInfo> getTaskInfo()
             throws RuntimeException
     {
-        if (scheduledFuture != null && scheduledFuture.isCancelled()) {
+        if (scheduledFuture != null && scheduledFuture.isCancelled() && lastException.get() != null) {
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:147]: What is this change about? Seems unrelated.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcess.java [hunk 2]
```diff
@@ -182,6 +182,10 @@ public int getPort()
         return port;
     }
 
+    public URI getLocation()
+    {
+        return location;
+    }
     private static URI getBaseUriWithPort(URI baseUri, int port)
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcess.java:189]: add empty line

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcessFactory.java [hunk 3]
```diff
@@ -38,10 +38,8 @@
 
 public class NativeExecutionProcessFactory
 {
-    // TODO add config
-    private static final int MAX_THREADS = 1000;
     private static final Duration MAX_ERROR_DURATION = new Duration(2, TimeUnit.MINUTES);
-
+    public static final URI DEFAULT_URI = URI.create("http://127.0.0.1");
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcessFactory.java:42]: Should this not include a port?

File: presto-spark-base/src/test/java/com/facebook/presto/spark/PrestoSparkQueryRunner.java [hunk 4]
```diff
@@ -405,9 +405,9 @@ public PrestoSparkQueryRunner(
     private static void setupLogging()
     {
         Logging logging = Logging.initialize();
-        logging.setLevel("org.apache.spark", INFO);
+        logging.setLevel("org.apache.spark", WARN);
```
Comment (mbasmanova) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/PrestoSparkQueryRunner.java:408]: Can you explain this change? Seems unrelated. If possible, it would be great to extract unrelated changes into a separate PR.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+import static java.util.Objects.requireNonNull;
+
+/**
+ * PrestoSparkNativeTaskExecutorFactory is responsible for launching the external native process and managing the communication
+ * between Java process and native process (by using the {@Link NativeExecutionTask}).
+ * It will send necessary metadata (e.g, plan fragment, session properties etc.) as a part of
+ * BatchTaskUpdateRequest. It will poll the remote CPP task for status and results (pages/data if applicable)
+ * and send these back to the Spark's RDD api
+ *
+ * PrestoSparkNativeTaskExecutorFactory is singleton instantiated once per executor
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:101]: nit: add period at the end of the sentence.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+ *
+ * For every task it receives, it does the following
+ * 1. Create the Native execution Process (NativeTaskExecutionFactory) ensure that is it created only once.
+ * 2. Serialize and pass the planFragment, source-metadata (taskSources), sink-metadata (tableWriteInfo or shuffleWriteInfo)
+ *    and submit a nativeExecutionTask.
+ * 3. Return Iterator to spark who next methods will
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:107]: typo: who

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+ *    3.a Call {@link NativeExecutionTask}'s pollResult() to retrieve {@link SerializedPage} back from external process.
+ *    3.b If not more output is available, then check if task has finished successfully or with exception
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:109]: not -> no

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.facebook.presto.sql.planner.SchedulingOrderVisitor.scheduleOrder;
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher.searchFrom;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.units.DataSize.succinctBytes;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * PrestoSparkNativeTaskExecutorFactory is responsible for launching the external native process and managing the communication
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:95]: Thank you for writing this detailed comment.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+
+        Session session = taskDescriptor.getSession().toSession(
+                sessionPropertyManager,
+                taskDescriptor.getExtraCredentials(),
+                extraAuthenticators.build());
+        PlanFragment fragment = taskDescriptor.getFragment();
+        StageId stageId = new StageId(session.getQueryId(), fragment.getId().getId());
+        TaskId taskId = new TaskId(new StageExecutionId(stageId, 0), partitionId, attemptNumber);
+
+        // TODO: Remove this once we can display the plan on Spark UI.
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:212]: This is quite unfortunate. But, why do we need every executor to log plan fragment? Shouldn't we have driver log these once?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+        // Currently, `textPlanFragment` throws an exception if json-based UDFs are used in the query, which can only
+        // happen in native execution mode. To resolve this error, `JsonFileBasedFunctionNamespaceManager` must be
+        // loaded on the executors as well (which is actually not required for native execution). To do so, we need a
+        // mechanism to ship the JSON file containing the UDF metadata to workers, which does not exist as of today.
+        // TODO: Address this issue; more details in https://github.com/prestodb/presto/issues/19600
+        log.info("Logging plan fragment is not supported for presto-on-spark native execution, yet");
+
+        if (fragment.getPartitioning().isCoordinatorOnly()) {
+            throw new UnsupportedOperationException("Co-ordinatory only fragment execution is not supported by native task executor");
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:221]: typo: Co-ordinatory

**Review by mbasmanova — COMMENTED** [review]
@shrinidhijoshi Love this change. Skimmed the first commit and make some minor comments. Overall, I find this change well written with nice documentation and easy to follow code. Thank you for taking the time to put this in a such beautiful shape.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java [hunk 11]
```diff
@@ -179,35 +177,6 @@ private SubPlan buildFragment(PlanNode root, FragmentProperties properties, Plan
                     tableWriterNodeIds);
         }
 
-        // Only delegate non-coordinatorOnly plan fragment to native engine
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java:182]: Awesome!

**Review by mbasmanova — APPROVED** [review]
@shrinidhijoshi Very happy to see this change. Let's get this in.

File: presto-spark-base/src/test/java/com/facebook/presto/spark/PrestoSparkQueryRunner.java (hunk 4 above)
Comment (shrinidhijoshi) [reply presto-spark-base/src/test/java/com/facebook/presto/spark/PrestoSparkQueryRunner.java:408]: Will remove this. Its to prevent flooding logs in CI.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java (hunk 9 above)
Comment (shrinidhijoshi) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:212]: PlanFragment log in executor aids debugging to easily spot which fragment is being executed. Also, given internally at Meta, executors logs always contains logs of ALL the tasks run on that executor v/s just that task, it is hard to visually draw boundaries on these logs without plan fragment logging.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcessFactory.java (hunk 3 above)
Comment (shrinidhijoshi) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcessFactory.java:42]: The port is selected dynamically by the `NativeExecutionProcess`/`DetachedNativeExecutionProcess` so we don't need to specify it here.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java (hunk 1 above)
Comment (shrinidhijoshi) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:148]: While this class was being used by `NativeExecutionOperator`, it was assumed that `getTaskInfo` will always be called while task is running. 
The way the new `TaskExecutorFactory` is implemented, that is not the case. We `stop()` the task first and then do a final query to get `TaskInfo` for statistics.

I can change that implementation to extract `TaskInfo` before stopping, but it seemed like improving this check was a better approach. 

Not leaning strongly either way, let me know if you have a preference
